academic_observatory_workflows/workflows/crossref_events_telescope.py

```diff
@@ -12,383 +12,859 @@
+        # we don't actually use this data
+        if next_cursor:
+            with open(cursor_path, "a") as f:
+                f.write(next_cursor + "\n")
+
+        # Write events to file
+        if events:
+            total_events += len(events)
+            with open(data_path, "a") as f:
+                with jsonlines.Writer(f) as writer:
+                    writer.write_all(events)
+
+        # If next_cursor is None exit loop
+        if not next_cursor:
+            break
+
+    # Remove cursor as we have finished
+    if os.path.isfile(cursor_path):
+        os.remove(cursor_path)
+
+    logging.info(f"{request}: successfully downloaded {total_events} events")
+
+
+def fetch_events(request: EventRequest, cursor: str = None, n_rows: int = 1000) -> Tuple[Dict, str]:
+    """Fetch the events for an EventsRequest from the given url until no new cursor is returned or a RetryError occurs.
     The extracted events are appended to a jsonl file and the cursors are written to a text file.
 
-    :param url: The url
-    :param headers: The headers dict
-    :param events_path: Path to the file in which events are stored.
-    :param cursor_path: Path to the file where cursors are stored.
-    :return: next_cursor, counter of events and total number of events according to the response
+    :param request: the url.
+    :param cursor: the cursor.
+    :param n_rows: number of rows per page.
+    :return: the fetched events and the next_cursor.
     """
+
+    headers = {"User-Agent": get_user_agent(package_name="academic_observatory_workflows")}
     try:
         crossref_events_limiter()
+        url = request.make_url(rows=n_rows, cursor=cursor)
         response = retry_get_url(url, headers=headers)
     except requests.exceptions.RequestException:
-        # Try again with rows set to 100
-        url = re.sub("rows=[0-9]*", "rows=100", url)
         crossref_events_limiter()
+        url = request.make_url(rows=100, cursor=cursor)
         response = retry_get_url(url, headers=headers)
 
     response_json = response.json()
-    total_events = response_json["message"]["total-results"]
     events = response_json["message"]["events"]
     next_cursor = response_json["message"]["next-cursor"]
-    counter = len(events)
 
-    # append events and cursor
-    if events:
-        with open(events_path, "a") as f:
-            with jsonlines.Writer(f) as writer:
-                writer.write_all(events)
-    if next_cursor:
-        with open(cursor_path, "a") as f:
-            f.write(next_cursor + "\n")
-    return next_cursor, counter, total_events
+    return events, next_cursor
+
+
+def crossref_events_limiter(calls_per_second: int = 10):
+    """Function to throttle the calls to the Crossref Events API"""
+
+    identifier = "crossref_events_limiter"
+    item = RateLimitItemPerSecond(calls_per_second)  # 10 per second
+
+    while True:
+        if not moving_window.test(item, identifier):
```

Comment:
Should moving_window be capitalised to indicate a global variable?

Reply:
Yeah that makes sense.